### DISCUSSION TITLE
Controlling nested machine states from parent, anti-pattern ?

### DISCUSSION THREAD

**Discussion by blaadje** [statelyai/xstate#2121] [Q&A]
Hey! I'm new with final state machine and wonder if my approach is relevant. _(Btw I love xstate, you can really do a lot & the documentation is really good, thanks for that! ❤️ )_

So I need to make **api calls** in parallel when initialising my app. 
I want each **api call to have its own machine** so I can retrieve each state (IDLE, LOADING, SUCCESS, FAILURE) for one call & use the machine everywhere in the app. But I also want to wrap these machines in a parent one which will also have the same states depending on the submachines & the ability to trigger all calls.

I tried with 
- dedicated machine for doing the fetch (used as a service) having states depending of the sendParent. However this machine can't listen to the child sendParent as the spawning doesnt happen here
- actions but its not async
- lots of others things which didn't lead anywere 😅 

---

So here the idea

Child machine :

```js
export default Machine({
  initial: IDLE,
  context: { data: [], error: '' },
  states: {
    [IDLE]: {
      on: {
        FETCH: { target: LOADING },
      },
    },
    [LOADING]: {
      invoke: {
        src: getData,
        onDone: {
          target: SUCCESS,
          actions: assign({ data: (_, { data }) => data.data }),
        },
        onError: {
          target: FAILURE,
          actions: assign({ error: (_, { data }) => data.data }),
        },
      },
    },
    [SUCCESS]: {
      entry: sendParent('CALL_MACHINE1_SUCCESS'),
      type: 'final',
    },
    [FAILURE]: {
      entry: sendParent('CALL_MACHINE1_FAILURE'),
    },
  },
})
```

Parent machine :
This code is not working and would not be the best option but that's basically the idea behind.

```js
export default Machine({
  initial: IDLE,
  context: {
    callMachine1Ref: null,
    callMachine2Ref: null,
  },
  entry: assign({
    callMachine1: () => spawn(callMachine1, { sync: true }),
    callMachine2: () => spawn(callMachine2, { sync: true }),
  }),
  states: {
    [IDLE]: {
      on: {
        FETCH: {
          target: LOADING,
          actions: [
            // fetch here both child machines somehow 
            (context, { id }) =>
              context.simulationRef.send({ type: 'FETCH', id }),
            (context, { id }) => context.recipeRef.send({ type: 'FETCH', id }),
          ],
        },
      },
    },
    [LOADING]: {
      on: {
        // here I need something to check that both events sent above are successful
        [SUCCESS]: {},
      },
    },
    [SUCCESS]: {},
    [FAILURE]: {},
  },
})
```

By doing so, my gain is that in my vue component using the parentMachine I don't need to :
- import/interpret 2 (or more) machines
- send multiple event
- look for multiple states to know when both are successful

---

I hope my problematic is clear. So my questions are :
- Is this relevant to have such approach ?
- Can we have something similar ?

**Comment by Andarist** [comment]
> here I need something to check that both events sent above are successful

You can achieve that with `always` transitions.

> fetch here both child machines somehow 

You can redispatch a command event to child machines by using `pure` action, smth like:
```js
pure(ctx => [callMachine1, callMachine2].map(child => send({ type: 'FETCH' }, { to: child })))
```

--- 

It all kinda depends on what you are trying to achieve. It makes sense for child machines to report their status to the parent and for the parent to determine the collective state - a file uploader would also be a good example of that. You can have multiple uploads in parallel but you really don't want the send button to be clickable when there are my uploads in progress.

If the list of child machines is not dynamic you could model them as parallel states with invokes and leverage `onDone` on the parallel state to transition to other states.

**Reply by blaadje** [reply]
@Andarist 

> If the list of child machines is not dynamic you could model them as parallel states with invokes and leverage onDone on the parallel state to transition to other states.

Thats what I tried to do. 
- How would you do that by keeping parent machine states simple (IDLE, LOADING, SUCESS, FAILURE), no LOADING.callMachine1 substates.
- How would you do without spawning a machine in the `invoke.src` as its already spawned in entry
- How would you retrieve the SUCCESS state for each machine.
By reading this https://xstate.js.org/docs/guides/communication.html#invoking-machines it doesnt seem to fit my needs.

Could you give me example so I can fully understand how achieve that ?

> You can redispatch a command event to child machines by using pure action

Alright I see. Just to make sure, will those actions will be in parallel ?
